docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,121 @@
+---
+title: "TypeScript Clients Across Wing Applications"
+description: TypeScript Clients Across Wing Applications
+---
+
+# TypeScript Clients Across Wing Applications
+
+- **Author(s)**: @MarkMcCulloh
+- **Submission Date**: 2024-06-02
+
+## Summary
+
+Allow for the generation of TypeScript clients that can interact with wing-defined infrastructure.
+Unlike `extern` or `@inflight`, the client is meant to be consumed by TypeScript applications rather than directly executed by Wing itself.
+
+The goal of this document is to walk through the DX of this experience.
+
+An important consideration for this discussion is allowing the generated client to reference resources defined from a separate wing application.
+To facilitate this separation, it seems necessary to introduce a mechanism to export/import deployed data between Wing apps.
+The rest of this proposal does it via `cloud.Outputs`, but that likely deserves its own RFC.
+
+## Example Situation
+
+Acme Co. has a platform team and several application teams. Their platform team manages Kubernetes clusters in AWS with additional AWS resources as needed.
+Application teams write applications in TypeScript and deploy them to these shared Kubernetes clusters.
+The centralized infrastructure definitions are written in Wing in their own repository.
+The applications are written in TypeScript but also have Wing files for infrastructure/deployment.
+
+To add a new application, the platform team needs to ensure that the application has the necessary permissions to interact with the shared resources. To do this, a kubernetes namespace and service account is created:
+
+```wing
+let namespace = cluster.createNamespace("app123");
```

Comment:
Let's assume users don't need wing to define k8s resources

Reply:
They won't use wing to define k8s resources on either the core infra or the application side? So they would only use wing to define other resources?

Wing could be used to create e.g. an IAM role, and then the platform team (or app team) could manually reference this role in their manifests. This wouldn't the RFC overall, to me that would still necessitate the `cloud.Outputs` to somehow get this information out.

Reply:
Yes, permissions are platform-specific and definitely require some solution in the RFC. 

Maybe when `TypeScriptClient` is compiled to AWS it would provision an IAM policy resource of some kind so that the app's role could use.

Maybe the user will be required in this case to specify the name of the IAM role/policy to provision in `wing.toml` so that the app can reference.